Kick sim

kick-sim opened the pull request:
## 요약 
**Controller**

- 사용자의 요청을 처리하는 역할
- CategoryController - category 관련 요청 처리
- ProductController - product 관련 요청 처리

**Dto**

- request 폴더 - 요청에 필요한 필드를 담은 클래스 구현
- response 폴더 - 응답에 필요한 필드를 담은 클래스 구현

**Entity**

- Category와 Product라는 테이블의 역할을 하는 클래스를 담은 스키마 역할

**Repository**

- 데이터베이스 역할을 하는 클래스를 담는 역할
- 데이터베이스 관리하는 SQL문 역할

**Service**

- 사용자의 요청에 맞도록 정보를 가공하는 역할
- CategoryService - category 관련 요청 정보 가공
- ProductService - product 관련 요청 정보 가공

## 어려웠던 점

- MVC 패턴으로 개발해본 경험이 부족해서 실행 과정을 이해하는 것이 어려웠습니다.

## 코드리뷰 받고싶은 점

- Entity에 만들어둔 Product 클래스에 변수가 정해져 있는데 이번 과제 요구사항에 보면 상품 목록을 출력할 때는 id, categoryId, name, createdAt을 Products 객체에 담아서 응답을 보내야하고 상품 세부사항을 출력할 때는 id, categoryId, categoryName, name, description, createdAt Product객체에 담아서 응답해야해서 두 응답을 위한 Product정보가 달라서 서로 다른 객체를 쓰려고 하니 목록을 출력할때 DB에 담긴 Product 정보 중 description을 빼고 새로운 DB를 만들어서 보내는 방법만 떠올랐는데 너무 비효율적인 작업이라 생각되었습니다. 이 부분을 해결하려면 어떻게 해야하는지 궁금합니다.

kick-sim committed:
Implementaion Reservation System

kick-sim committed:
Implementation Reservation System

junghunkang commented on src/main/java/com/landvibe/summer/reservation/controller/CategoryController.java:
```diff
@@ -0,0 +1,35 @@
+public class CategoryController {
+    private final CategoryService categoryService;
+
+    @RequestMapping(value = "/api/category", method = RequestMethod.POST)
```
매서드에 붙이는 @requestmapping -> @getmapping으로 사용하는게 좋습니다
https://change-words.tistory.com/entry/Spring-%EC%8A%A4%ED%94%84%EB%A7%81-RequestMapping-%EB%8C%80%EC%8B%A0-PostMapping-GetMapping-%EC%93%B0%EB%8A%94-%EC%9D%B4%EC%9C%A0
관련해서 이 문서를 한번 읽어보면 도움이 될듯 합니당 ~

junghunkang replied:
추가적으로 하나의 Controller에 공통된 url 은 Class를 명시하는 시점에 @requestMapping을 사용하여 공통되어 사용할 수 있습니다
한번 찾아서 바꿔보는거 추천~

junghunkang commented on src/main/java/com/landvibe/summer/reservation/controller/CategoryController.java:
```diff
@@ -0,0 +1,35 @@
+
+    @RequestMapping(value = "/api/categories", method = RequestMethod.GET)
+    public CategoriesResponse lookUp() {
+        ArrayList<Category> categories = categoryService.lookUp();
```
ArrayList의 타입으로 필드를 정의하였는데
필드의 타입은 List (인터페이스)를 사용하는게  더 좋을것 같습니다
대부분의 실제 객체는 ArrayList를 많이 사용하지만 LinkedList 등의 여러 타입이 반환되어도
확성장의 측면에서 더 유용할듯 싶습니다

junghunkang commented on src/main/java/com/landvibe/summer/reservation/controller/ProductController.java:
```diff
@@ -0,0 +1,38 @@
+public class ProductController {
+    private final ProductService productService;
+
+    @RequestMapping(value = "/api/product", method = RequestMethod.POST)
```
이것도 @postmapping 어노테이션을 사용하는게 더 좋을듯 합니당
관련해서 위에 전달한 링크를 보면 좋을듯

junghunkang commented on src/main/java/com/landvibe/summer/reservation/service/CategoryService.java:
```diff
@@ -0,0 +1,46 @@
+    private final CategoryRepository categoryRepository;
+
+    public Long join(CategoryRequest request) {
+        if (!validateProduct(request)) {
```
유효성을 검사한뒤 결과를 알려주는 방법에는 여러가지가 있겠지만
Custom Exception을 구현하고 예외를 발생시키는 방법도 있습니당 (스터디에서는 그냥 정의된 예외를 발생시켰지만..)
https://velog.io/@banjjoknim/RestControllerAdvice
관련해서 발생한 예외를 한곳에서 공통으로 처리할 수 있는데 한번 읽어보면 좋을듯

junghunkang commented on src/main/java/com/landvibe/summer/reservation/service/ProductService.java:
```diff
@@ -0,0 +1,52 @@
+        if (!validateProduct(request)) {
+            return -1L;
+        }
+        Product newProduct = new Product(request.getCategoryId(), request.getName(), request.getDescription(), null, LocalDateTime.now());
```
한 매서드, 생성자에 파라미터가 많아지면 코드의 가독성(어떤 자리에 어떤값이 들어가는지 한눈에 알아보기가 힘들다)이 조금 떨어지는.. 영향이 있는것 같습니다
관련해서 파라미터를 하나의 클래스에 묶어두거나 Builder를 사용하는 방법도 있는데
찾아서 바꿔보면 좋을듯

junghunkang commented on src/main/resources/templates/home.html:
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
```
👍 
뷰도 한번 써봤넴

junghunkang reviewed (commented):
하나의 Entity? Domain 당 하나의 Request, Response DTO를 둘 필요는 없습니당
응답 필요에 따라서 여러개를 생성해서 반환해주는 방법으로 고민을 해결할수 있지 않을까 생각이 드네염
ResponseEntity라는 Wrapping Class를 지원하는데 이것도 구글링해서 한번 알아보면 좋을듯~

kick-sim committed:
RequestMapping->Get/PostMapping으로 변경
공통된 URL Class명시 시점 @RequestMapping으로 공통으로 사용
파라미터가 많아 떨어진 가독성 향상을 위해 Builder 사용

kick-sim committed:
들여쓰기 수정
